### DISCUSSION TITLE
Proposal: Pluggable route variables

### DISCUSSION THREAD

**Discussion by philproctor** [ThreeMammals/Ocelot#1903] [Ideas]
## New Feature
Downstream route variables could pull from a pluggable architecture. I have implemented a similar mechanism in our companies usage of Ocelot, where downstream variables could be of the format {claim->UserID} to replace a variable with the first UserID claim and similar for request headers, cookies, and other variables. It also allows operations such as concatenating variables or similar. We also use it in claim authorization for routes.

## Motivation for New Feature
All downstream variables currently require the variable to be a replacement from the upstream request. A system such as this would significantly increase flexibility allowing things such as routing to a URL based on the currently logged in user.

### Update 1 by Maintainer on Jan 7, 2024
- [Claims Transformation](https://ocelot.readthedocs.io/en/latest/features/claimstransformation.html) feature

**Comment by TomPallister** [comment]
@philproctor this is a great feature.

My only comment is that Ocelot already has a bunch of request manipulation / transform capabilities..and these are all a bit disperate and hacky. It might be a good idea to pull them all together so the code is clear and concise but maybe that is a larger refactor than what you are planning.

Also make sure no acceptance tests break as I guess you will deffo be breaking unit tests that will need to change.

**Comment by raman-m** [comment]
@philproctor 
> We also use it in claim authorization for routes.

It seems you override [AuthorizationMiddleware](https://github.com/ThreeMammals/Ocelot/blob/develop/src/Ocelot/Middleware/OcelotPipelineConfiguration.cs#L48) to process and manipulate claims, right?

So, we didn't see your PR or more detailed description here...
I understood that your feature request is very rare user case for your specific needs only. Am I right?
Could you show your idea better in details plz? Any sample code?

**Comment by raman-m** [comment]
> #### Motivation for New Feature
> All downstream variables currently require the variable to be a replacement from the upstream request. A system such as this would significantly increase flexibility **allowing things such as routing to a URL based on the currently logged in user**.

In this scenario Ocelot must authenticate and make a redirection, right?
Doesn't "**RouteClaimsRequirement**" property work well? You've authorized with a claim, then you have a request to downstream host.
Just specify multiple routes for each user.

And, what store will be for data like **user** and **URL**?
You can pack and create special tokens with all required data/claims.

In my opinion, all such transformations can be done by custom middleware, or even delegating handler.
In that dev case you'll have one single route in JSON. 
So, this proposal design is unclear and it should be described in details.
